### PR TITLE
Fix menu dropdown animation

### LINKED ISSUE #756: [Global] Animation persists when reduce motion is enabled
## Issue summary

Through out Dawn theme there are various, subtle animation. Ex., navigation drop-down, product card hover, blog post hover, etc. These animations persist when the user has set their [operating system accessibility settings to reduce motion](https://www.smashingmagazine.com/2020/09/design-reduced-motion-sensitivities/#reduced-motion-from-the-users-perspective).

Current animations could cause issues for folks with a Vestibular disorder or anyone susceptible to motion sickness, specifically the viewport-wide sweeping motion when a blog post item is moused over.

<details>
  <summary>Screenshots</summary>

  Gif animation of Dawn blog list page. A mouse cursor hovers over each item showcasing a viewport-wide zoom animation of the blog post image.
</details>

#### CSS

```css
.article-content:hover img {
  transform: scale(1.07);
}
.article-content img {
  transition: transform var(--duration-long) ease;
}

/* … */

.card-wrapper:hover .media.media--hover-effect > img + img {
  /* … */
  transition: transform var(--duration-long) ease;
  transform: scale(1.03);
}

/* … */

.media > img {
  /* … */
  transition: opacity 0.4s cubic-bezier(0.25, 0.46, 0.45, 0.94);
}

/* … */

.header__inline-menu details[open] > .header__submenu {
  /* … */
  transform: translateY(0);
  animation: animateMenuOpen var(--duration-default) ease;
  /* … */
}
```

## Steps to reproduce

1. Go to a Dawn theme store with a password enabled with any supported browser
2. Navigate to the Blog listing page
3. Mouse-over blog items with images

### Expected

- Animation to be lowered or cease completely with reduced motion setting enabled.

### Actual

- Animations persist.

## Recommendation

**Up for discussion.**

Adjust the `transition`/`transform` values to lessen the animation (or consider removing completely) when the reduce motion operating system setting is enabled.

Use the `prefers-reduced-motion` CSS media query to achieve the desired affect.

#### CSS

```css
@media screen and (prefers-reduced-motion) {
  .article-content:hover img {
    transform: none;
  }
  .article-content img {
    transition: none;
  }
}
```

<!-- #### Diff

```diff
- line removed
+ line added
``` -->

## Specifications

- **Browser:** Any
- **Dawn version:** 2.2.0
- **Component:** Global
- **WCAG Principal:** Operable
- **WCAG SC:** [2.3.3 Animation from Interactions](https://www.w3.org/WAI/WCAG21/Understanding/animation-from-interactions.html)
- **Severity:** Medium
- **Effort:** Low

### LINKED ISSUE #1287: Menu dropdown doesn't animation consistently
**Describe the current behavior**

- Open desktop menu dropdown
- Animation plays when dropdown is opened for first time
- Animation does not play consistently on subsequent opening of dropdown

**Describe the expected behavior**

- Animation plays consistently every time dropdown is opened

**Version information (Dawn, browsers and operating systems)**
- Dawn Version: 3.0.0
- Chrome Version 97.0.4692.99
- macOS Version 12.1

**Possible solution**

n/a

**Additional context/screenshots**

screencast 2022-01-31 10-00-06

### PR DESCRIPTION
**Pull request by KaichenWang**
## **Why are these changes introduced?**

1. Fixes #1287
2. Partially fixes #756

### 1. Issue 1287

- Continue to leverage animations defined in CSS
- Add JS enhancement by using Web Animations API (WAAPI) to control animation playback
- When menu `details` is toggled for first time, cache CSS animations in JS
- Use WAAPI animate `play()` and `cancel()` methods when menu is subsequently toggled

### 2. Issue 756

- Leverage existing `.motion-reduce` class: set `animation` CSS property to `none`

### Alternatives
There's a CSS-only method that can be used on native details/summary elements to reset animation, but this doesn't work in Dawn, because we've already enhanced native behaviour with JS (for ensuring only 1 menu item is open at a time, a11y, etc.)
``` css
summary:active + * {
  animation-name: none;
}
```

### Browser compatibility 
Works in all supported browsers
Screen Shot 2022-01-31 at 10 28 21 AM

### Performance 

There should be no noticeable performance impact

> CSS-based animations, and Web Animations where supported natively, are typically handled on a thread known as the "compositor thread". This is different from the browser's "main thread", where styling, layout, painting, and JavaScript are executed. This means that if the browser is running some expensive tasks on the main thread, these animations can keep going without being interrupted.

Source: https://developers.google.com/web/fundamentals/design-and-ux/animations/animations-and-performance

## **Demo links**

- [Store](https://os2-demo.myshopify.com/?preview_theme_id=127525978134)
- [Editor](https://os2-demo.myshopify.com/admin/themes/127525978134/editor)

## **Checklist**
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

**Review by melissaperreault — COMMENTED** [review]
Thank you Kai!  

- [x] From what I was able to test, I noticed the animation jumps when navigating fast and slowly. ([Video reference](https://screenshot.click/22-01-zmb1h-y6je0.mp4))

**Comment by KaichenWang** [comment]
@melissaperreault Thanks for the feedback! Updated the animation to prevent the flicker  (https://github.com/Shopify/dawn/pull/1291/commits/5701000f65b222c5f42f57894143986cacac1ca4). Let me know if this fixes the issue. Sharing some slo-mo footage below to understand what was happening. 

### Before

https://user-images.githubusercontent.com/3710548/151999340-eb1cf4ee-0a41-4c1a-a3ad-cd2f3ff485ee.mp4

### After

https://user-images.githubusercontent.com/3710548/151999331-cf36d360-2959-4010-bc13-9ac8786b5f50.mp4

**Comment by KaichenWang** [comment]
> The issue seem to be only in Safari originally right ? I wasn't able to replicate in chrome.

Also an issue on Chrome. May work for first few clicks, but eventually animation doesn't play consistently anymore. Easier to visualize when animations are set to 25% speed in dev tools.

File: assets/base.css [hunk 1]
```diff
@@ -2451,10 +2452,14 @@ details[open] > .header__icon--menu .icon-hamburger {
   transform: scale(1.07);
 }
 
-.header__inline-menu details[open] > .header__submenu {
-  opacity: 1;
-  transform: translateY(0);
+.header__inline-menu details > .header__submenu {
+  opacity: 0;
+  transform: translateY(-1.5rem);
```
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/base.css:2457]: In my testing, flicker may still occur occasionally without this. 

My theory is that if we don't define this explicitly then default values are:
``` css
opacity: 1;
transform: none;
```
The animation is trying to start by setting:
``` css
opacity: 0;
transform: translateY(-1.5rem);
```
I think it's essentially a race condition that determines which set of styles is applied first. If the default is applied first, the user will observe flickering (see https://github.com/Shopify/dawn/pull/1291#issuecomment-1026982784)

**Review by ludoboludo — DISMISSED** [review]

**Comment by KaichenWang** [comment]
> I noticed that when refreshing the page, the first time you swap between the menus you will get the flicker in Chrome. All of the subsequent animations work properly. In this video, the first few clicks are showing pre-refresh, and then the last click is after refreshing the page and you can see the flicker.
> 
> It seems to be 100% reproducible, but does require a user's first interaction with the menu to be a pretty quick opening of one and then the other - if there's any sort of delay more than a second or two then the flicker doesn't seem to happen.
> 
>  Screen.Recording.2022-02-01.at.3.18.43.PM.mov

I see definitely the flicker in your video, but I've not been able to reproduce on my end. May need to sync on what factors could causing this.

**Comment by KaichenWang** [comment]
> _Note for thoughts: I pushed the thinking further with more attention and was wondering if the animation could also apply inside the dropdown when opening the third level. The short answer is no as users are on an active task and that it might distract them from completing their task efficiently but I'd love us (UX) to think about these further more to strengthen our animation principle. ([initial thought video reference](https://screenshot.click/22-02-2qfov-jshzd.mp4))_

@melissaperreault The dropdown for level 3 links actually have an animation defined, but it suffers from the same problem as the top level animation. It plays once, then it becomes inconsistent (see GIF below). I haven't audited the entire theme, but I suspect there are more cases of this. I think we can use the techniques from this PR to fix other instances we identify in a follow up issue.

screencast 2022-02-02 10-43-26

**Review by tyleralsbury — APPROVED** [review]
Refresh bug is fixed - thanks Kai!

**Pull request merged**